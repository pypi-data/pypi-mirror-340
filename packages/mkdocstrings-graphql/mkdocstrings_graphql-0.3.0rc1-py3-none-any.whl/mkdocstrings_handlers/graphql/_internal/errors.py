from __future__ import annotations

import sys
from typing import TYPE_CHECKING, final

if TYPE_CHECKING:
    from pathlib import Path

if sys.version_info >= (3, 12):
    from typing import override
else:
    from typing_extensions import override  # pyright:ignore[reportUnreachable]


@final
class GraphQLFileSyntaxError(Exception):
    def __init__(self, path: str | Path, message: str):
        super().__init__()
        self.message = self.format_message(path, message)

    @override
    def __str__(self) -> str:
        return self.message

    def format_message(self, path: str | Path, message: str) -> str:
        """Builds final error message from path to schema file and error
        message.

        Args:
            path: A `str` or `PathLike` object pointing to a file that failed
                to validate.
            message: A `str` with validation message.

        Returns:
            Final error message.
        """
        return f"Could not load {path}:\n{message}"
