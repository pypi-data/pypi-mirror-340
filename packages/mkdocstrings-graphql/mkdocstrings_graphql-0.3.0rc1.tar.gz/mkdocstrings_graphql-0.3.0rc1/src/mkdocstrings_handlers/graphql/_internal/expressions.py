from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class Expr:
    canonical_path: str | None = field(init=False)

    def __post_init__(self) -> None:
        self.canonical_path = None

    @property
    def class_name(self) -> str:
        return self.__class__.__name__


@dataclass
class TypeName(Expr):
    name: str

    @property
    def render(self) -> str:
        return self.name


@dataclass
class Annotation(Expr):
    name: str
    non_null: bool
    is_list: bool
    non_null_list: bool

    @property
    def render(self) -> str:
        rendered = self.name
        if self.non_null:
            rendered = f"{rendered}!"
        if self.is_list:
            rendered = f"[{rendered}]"
        if self.non_null_list:
            rendered = f"{rendered}!"
        return rendered
