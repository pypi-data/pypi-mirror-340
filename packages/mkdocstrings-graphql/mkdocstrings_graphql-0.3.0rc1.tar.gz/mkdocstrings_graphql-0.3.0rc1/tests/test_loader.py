from dataclasses import asdict
from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from mkdocstrings_handlers.graphql._internal.errors import GraphQLFileSyntaxError
from mkdocstrings_handlers.graphql._internal.expressions import Annotation, TypeName
from mkdocstrings_handlers.graphql._internal.loader import Loader
from mkdocstrings_handlers.graphql._internal.models import (
    EnumTypeNode,
    EnumValue,
    Field,
    Input,
    InputObjectTypeNode,
    InterfaceTypeNode,
    ObjectTypeNode,
    OperationTypeNode,
    ScalarTypeNode,
    UnionTypeNode,
)


def test_graphql_syntax_error(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            type TestObject {
                field1: String!
                field2: String!
            """)

    with pytest.raises(GraphQLFileSyntaxError) as excinfo:
        Loader(schema_paths=[tmp_path])
    assert str(excinfo.value).startswith(f"Could not load {schema_path}")


def test_load_enum(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            enum TestEnum {
                ENUM_VALUE_1
                ENUM_VALUE_2
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestEnum"] == EnumTypeNode(
        name="TestEnum",
        path="schemaName.TestEnum",
        description="",
        values=[EnumValue(name="ENUM_VALUE_1", description=""), EnumValue(name="ENUM_VALUE_2", description="")],
    )


def test_load_input(tmp_path: Path) -> None:
    with open(tmp_path / "schema.graphql", "w") as f:
        f.write("""
            input TestInput {
                field1: String!
                field2: String!
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestInput"] == InputObjectTypeNode(
        name="TestInput",
        path="schemaName.TestInput",
        description="",
        fields=[
            Input(
                name="field1",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
            Input(
                name="field2",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
        ],
    )


def test_load_interface(tmp_path: Path) -> None:
    with open(tmp_path / "schema.graphql", "w") as f:
        f.write("""
            interface TestInterface {
                field1: String!
                field2: String!
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestInterface"] == InterfaceTypeNode(
        name="TestInterface",
        path="schemaName.TestInterface",
        description="",
        fields=[
            Field(
                name="field1",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
            Field(
                name="field2",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
        ],
        interfaces=[],
    )


def test_load_object(tmp_path: Path) -> None:
    with open(tmp_path / "schema.graphql", "w") as f:
        f.write("""
            type TestObject {
                field1: String!
                field2: String!
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestObject"] == ObjectTypeNode(
        name="TestObject",
        path="schemaName.TestObject",
        description="",
        fields=[
            Field(
                name="field1",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
            Field(
                name="field2",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
        ],
        interfaces=[],
    )


def test_load_operation(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            schema {
                mutation: Mutation
                query: Query
            }

            type Mutation {
                testMutation(
                    field1: String
                    field2: String
                ): Boolean
            }

            type Query {
                testQuery(
                    field1: String
                    field2: String
                ): Boolean
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["Mutation.testMutation"] == OperationTypeNode(
        name="testMutation",
        path="schemaName.Mutation.testMutation",
        description="",
        arguments=[
            Input(
                name="field1",
                description="",
                type=Annotation(name="String", non_null=False, is_list=False, non_null_list=False),
            ),
            Input(
                name="field2",
                description="",
                type=Annotation(name="String", non_null=False, is_list=False, non_null_list=False),
            ),
        ],
        type=Annotation(name="Boolean", non_null=False, is_list=False, non_null_list=False),
    )
    assert loader.schemas_collection["schemaName"]["Query.testQuery"] == OperationTypeNode(
        name="testQuery",
        path="schemaName.Query.testQuery",
        description="",
        arguments=[
            Input(
                name="field1",
                description="",
                type=Annotation(name="String", non_null=False, is_list=False, non_null_list=False),
            ),
            Input(
                name="field2",
                description="",
                type=Annotation(name="String", non_null=False, is_list=False, non_null_list=False),
            ),
        ],
        type=Annotation(name="Boolean", non_null=False, is_list=False, non_null_list=False),
    )


def test_load_scalar(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("scalar TestScalar")

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestScalar"] == ScalarTypeNode(
        name="TestScalar", path="schemaName.TestScalar", description=""
    )


def test_load_schema_definition(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            schema {
                mutation: Mutation
                query: Query
                subscription: Subscription
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"].definition is not None
    assert loader.schemas_collection["schemaName"].definition.mutation == "Mutation"
    assert loader.schemas_collection["schemaName"].definition.query == "Query"
    assert loader.schemas_collection["schemaName"].definition.subscription == "Subscription"
    assert loader.schemas_collection["schemaName"].definition.types == {"Mutation", "Query", "Subscription"}


def test_load_schema_definition_partial_definition(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            schema {
                mutation: Mutation
                query: Query
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"].definition is not None
    assert loader.schemas_collection["schemaName"].definition.mutation == "Mutation"
    assert loader.schemas_collection["schemaName"].definition.query == "Query"
    assert loader.schemas_collection["schemaName"].definition.subscription is None
    assert loader.schemas_collection["schemaName"].definition.types == {"Mutation", "Query"}


def test_load_union(tmp_path: Path) -> None:
    schema_path = tmp_path / "schema.graphql"
    with open(schema_path, "w") as f:
        f.write("""
            type TestObject1 {
                field: String
            }

            type TestObject2 {
                field: String
            }

            union TestUnion = TestObject1 | TestObject2
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert "TestObject1" in loader.schemas_collection["schemaName"]
    assert "TestObject2" in loader.schemas_collection["schemaName"]
    types = [TypeName(name="TestObject1"), TypeName("TestObject2")]
    types[0].canonical_path = "schemaName.TestObject1"
    types[1].canonical_path = "schemaName.TestObject2"
    assert loader.schemas_collection["schemaName"]["TestUnion"] == UnionTypeNode(
        name="TestUnion", path="schemaName.TestUnion", description="", types=types
    )


def test_parse_type(tmp_path: Path) -> None:
    with open(tmp_path / "schema.graphql", "w") as f:
        f.write("""
            type TestObject {
                nullElem: String
                nonNullElem: String!
                nullListNullElem: [String]
                nullListNonNullElem: [String!]
                nonNullListNullElem: [String]!
                nonNullListNonNullElem: [String!]!
            }
            """)

    loader = Loader(schema_paths=[tmp_path])
    loader.load(schema_name="schemaName")

    assert loader.schemas_collection["schemaName"]["TestObject"] == ObjectTypeNode(
        name="TestObject",
        path="schemaName.TestObject",
        description="",
        fields=[
            Field(
                name="nullElem",
                description="",
                type=Annotation(name="String", non_null=False, is_list=False, non_null_list=False),
            ),
            Field(
                name="nonNullElem",
                description="",
                type=Annotation(name="String", non_null=True, is_list=False, non_null_list=False),
            ),
            Field(
                name="nullListNullElem",
                description="",
                type=Annotation(name="String", non_null=False, is_list=True, non_null_list=False),
            ),
            Field(
                name="nullListNonNullElem",
                description="",
                type=Annotation(name="String", non_null=True, is_list=True, non_null_list=False),
            ),
            Field(
                name="nonNullListNullElem",
                description="",
                type=Annotation(name="String", non_null=False, is_list=True, non_null_list=True),
            ),
            Field(
                name="nonNullListNonNullElem",
                description="",
                type=Annotation(name="String", non_null=True, is_list=True, non_null_list=True),
            ),
        ],
        interfaces=[],
    )


def test_load_multiple_files(snapshot: SnapshotAssertion) -> None:
    schema_dir = Path(__file__).parent / "fixtures" / "schema"

    loader = Loader(schema_paths=[schema_dir / "constructs.graphql", schema_dir / "schema.graphql"])
    loader.load(schema_name="schemaName")

    collection_dict = asdict(loader.schemas_collection["schemaName"])
    assert collection_dict == snapshot
