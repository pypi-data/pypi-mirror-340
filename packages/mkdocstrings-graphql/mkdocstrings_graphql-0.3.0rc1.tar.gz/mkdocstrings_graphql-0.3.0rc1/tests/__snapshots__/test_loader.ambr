# serializer version: 1
# name: test_load_multiple_files
  dict({
    'definition': dict({
      'mutation': 'Mutation',
      'query': 'Query',
      'subscription': None,
      'types': frozenset({
        'Mutation',
        'Query',
      }),
    }),
    'kind': <Kind.SCHEMA: 'gql_schema'>,
    'members': dict({
      'Annotation': dict({
        'description': 'Interface containing all annotation combinations.',
        'fields': list([
          dict({
            'description': 'Nullable element.',
            'name': 'null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable element.',
            'name': 'non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with nullable elements.',
            'name': 'null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with non-nullable elements.',
            'name': 'null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable list with nullable elements.',
            'name': 'non_null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': 'Non-nullable list with non-nullable elements.',
            'name': 'non_null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'Annotation',
        'path': 'schemaName.Annotation',
      }),
      'AnnotationErrorImpl': dict({
        'description': 'Object implementing multiple interfaces.',
        'fields': list([
          dict({
            'description': 'Error message.',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable element.',
            'name': 'null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable element.',
            'name': 'non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with nullable elements.',
            'name': 'null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with non-nullable elements.',
            'name': 'null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable list with nullable elements.',
            'name': 'non_null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': 'Non-nullable list with non-nullable elements.',
            'name': 'non_null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': 'Custom element.',
            'name': 'custom',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'Int',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Annotation',
            'name': 'Annotation',
          }),
          dict({
            'canonical_path': 'schemaName.Error',
            'name': 'Error',
          }),
        ]),
        'name': 'AnnotationErrorImpl',
        'path': 'schemaName.AnnotationErrorImpl',
      }),
      'AnnotationImpl': dict({
        'description': '',
        'fields': list([
          dict({
            'description': 'Nullable element.',
            'name': 'null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable element.',
            'name': 'non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with nullable elements.',
            'name': 'null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Nullable list with non-nullable elements.',
            'name': 'null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Non-nullable list with nullable elements.',
            'name': 'non_null_list_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': False,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': 'Non-nullable list with non-nullable elements.',
            'name': 'non_null_list_non_null',
            'type': dict({
              'canonical_path': None,
              'is_list': True,
              'name': 'String',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': 'Custom element.',
            'name': 'custom',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'Int',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Annotation',
            'name': 'Annotation',
          }),
        ]),
        'name': 'AnnotationImpl',
        'path': 'schemaName.AnnotationImpl',
      }),
      'CreatePokemonError': dict({
        'description': '',
        'name': 'CreatePokemonError',
        'path': 'schemaName.CreatePokemonError',
        'types': list([
          dict({
            'canonical_path': 'schemaName.InvalidPokemonNumberError',
            'name': 'InvalidPokemonNumberError',
          }),
          dict({
            'canonical_path': 'schemaName.PokemonExistsError',
            'name': 'PokemonExistsError',
          }),
        ]),
      }),
      'CreatePokemonInput': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'name',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'types',
            'type': dict({
              'canonical_path': 'schemaName.Type',
              'is_list': True,
              'name': 'Type',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': '',
            'name': 'prevEvolutionNumbers',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': True,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'nextEvolutionNumbers',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': True,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'name': 'CreatePokemonInput',
        'path': 'schemaName.CreatePokemonInput',
      }),
      'CreatePokemonResponse': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'data',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNode',
              'is_list': False,
              'name': 'PokemonNode',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'error',
            'type': dict({
              'canonical_path': 'schemaName.CreatePokemonError',
              'is_list': False,
              'name': 'CreatePokemonError',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'CreatePokemonResponse',
        'path': 'schemaName.CreatePokemonResponse',
      }),
      'Error': dict({
        'description': 'Generic error interface.',
        'fields': list([
          dict({
            'description': 'Error message.',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'Error',
        'path': 'schemaName.Error',
      }),
      'ErrorInterface': dict({
        'description': 'Generic error interface.',
        'fields': list([
          dict({
            'description': 'Error message.',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'code',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'Int',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Error',
            'name': 'Error',
          }),
        ]),
        'name': 'ErrorInterface',
        'path': 'schemaName.ErrorInterface',
      }),
      'EvolutionNode': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'name',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'EvolutionNode',
        'path': 'schemaName.EvolutionNode',
      }),
      'InvalidPokemonNumberError': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Error',
            'name': 'Error',
          }),
        ]),
        'name': 'InvalidPokemonNumberError',
        'path': 'schemaName.InvalidPokemonNumberError',
      }),
      'Mutation.createPokemon': dict({
        'arguments': list([
          dict({
            'description': 'Parameters which determines to Pokemon to be created.',
            'name': 'input',
            'type': dict({
              'canonical_path': 'schemaName.CreatePokemonInput',
              'is_list': False,
              'name': 'CreatePokemonInput',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '''
              If we should the existing Pokemon if a duplicate with the same
              ``no`` exists.
            ''',
            'name': 'overwrite',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'Boolean',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
        ]),
        'description': '''
          Creates a Pokemon. If ``overwrite`` is true, overwrite the existing
          Pokemon if a duplicate with the same ``input.no`` exists.
        ''',
        'name': 'createPokemon',
        'path': 'schemaName.Mutation.createPokemon',
        'type': dict({
          'canonical_path': 'schemaName.CreatePokemonResponse',
          'is_list': False,
          'name': 'CreatePokemonResponse',
          'non_null': True,
          'non_null_list': False,
        }),
      }),
      'PokemonError': dict({
        'description': '',
        'name': 'PokemonError',
        'path': 'schemaName.PokemonError',
        'types': list([
          dict({
            'canonical_path': 'schemaName.InvalidPokemonNumberError',
            'name': 'InvalidPokemonNumberError',
          }),
          dict({
            'canonical_path': 'schemaName.PokemonNotFoundError',
            'name': 'PokemonNotFoundError',
          }),
        ]),
      }),
      'PokemonExistsError': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Error',
            'name': 'Error',
          }),
        ]),
        'name': 'PokemonExistsError',
        'path': 'schemaName.PokemonExistsError',
      }),
      'PokemonNode': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'name',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'types',
            'type': dict({
              'canonical_path': 'schemaName.TypeNode',
              'is_list': True,
              'name': 'TypeNode',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': '',
            'name': 'prevEvolutions',
            'type': dict({
              'canonical_path': 'schemaName.EvolutionNode',
              'is_list': True,
              'name': 'EvolutionNode',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
          dict({
            'description': '',
            'name': 'nextEvolutions',
            'type': dict({
              'canonical_path': 'schemaName.EvolutionNode',
              'is_list': True,
              'name': 'EvolutionNode',
              'non_null': True,
              'non_null_list': True,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'PokemonNode',
        'path': 'schemaName.PokemonNode',
      }),
      'PokemonNotFoundError': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'message',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'String',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
          dict({
            'canonical_path': 'schemaName.Error',
            'name': 'Error',
          }),
        ]),
        'name': 'PokemonNotFoundError',
        'path': 'schemaName.PokemonNotFoundError',
      }),
      'PokemonNumber': dict({
        'description': 'The Pokemon number, in the range of "0001" to "9999".',
        'name': 'PokemonNumber',
        'path': 'schemaName.PokemonNumber',
      }),
      'PokemonResponse': dict({
        'description': 'Response object for Query.pokemon.',
        'fields': list([
          dict({
            'description': 'The Pokemon specified by the query.',
            'name': 'data',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNode',
              'is_list': False,
              'name': 'PokemonNode',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': 'Errors encountered during the query.',
            'name': 'error',
            'type': dict({
              'canonical_path': 'schemaName.PokemonError',
              'is_list': False,
              'name': 'PokemonError',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'PokemonResponse',
        'path': 'schemaName.PokemonResponse',
      }),
      'Query.pokemon': dict({
        'arguments': list([
          dict({
            'description': "The Pokemon number, in the range of '0001' and '9999'.",
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'description': 'Queries for a Pokemon specified by ``no``.',
        'name': 'pokemon',
        'path': 'schemaName.Query.pokemon',
        'type': dict({
          'canonical_path': 'schemaName.PokemonResponse',
          'is_list': False,
          'name': 'PokemonResponse',
          'non_null': True,
          'non_null_list': False,
        }),
      }),
      'Query.weakness': dict({
        'arguments': list([
          dict({
            'description': 'The Pokemon number.',
            'name': 'no',
            'type': dict({
              'canonical_path': 'schemaName.PokemonNumber',
              'is_list': False,
              'name': 'PokemonNumber',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'description': 'Queries the weaknesses of the Pokemon specified by ``no``.',
        'name': 'weakness',
        'path': 'schemaName.Query.weakness',
        'type': dict({
          'canonical_path': 'schemaName.WeaknessResponse',
          'is_list': False,
          'name': 'WeaknessResponse',
          'non_null': True,
          'non_null_list': False,
        }),
      }),
      'Type': dict({
        'description': 'Generation I Pokemon types.',
        'name': 'Type',
        'path': 'schemaName.Type',
        'values': list([
          dict({
            'description': 'Bug type.',
            'name': 'BUG',
          }),
          dict({
            'description': 'Dragon type.',
            'name': 'DRAGON',
          }),
          dict({
            'description': 'Electric type.',
            'name': 'ELECTRIC',
          }),
          dict({
            'description': 'Fighting type.',
            'name': 'FIGHTING',
          }),
          dict({
            'description': 'Fire type.',
            'name': 'FIRE',
          }),
          dict({
            'description': 'Flying type.',
            'name': 'FLYING',
          }),
          dict({
            'description': 'Ghost type.',
            'name': 'GHOST',
          }),
          dict({
            'description': 'Grass type.',
            'name': 'GRASS',
          }),
          dict({
            'description': 'Ground type.',
            'name': 'GROUND',
          }),
          dict({
            'description': 'Ice type.',
            'name': 'ICE',
          }),
          dict({
            'description': 'Normal type.',
            'name': 'NORMAL',
          }),
          dict({
            'description': 'Poison type.',
            'name': 'POISON',
          }),
          dict({
            'description': 'Psychic type.',
            'name': 'PSYCHIC',
          }),
          dict({
            'description': 'Rock type.',
            'name': 'ROCK',
          }),
          dict({
            'description': 'Water type.',
            'name': 'WATER',
          }),
        ]),
      }),
      'TypeNode': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'id',
            'type': dict({
              'canonical_path': None,
              'is_list': False,
              'name': 'ID',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'type',
            'type': dict({
              'canonical_path': 'schemaName.Type',
              'is_list': False,
              'name': 'Type',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'TypeNode',
        'path': 'schemaName.TypeNode',
      }),
      'WeaknessError': dict({
        'description': '',
        'name': 'WeaknessError',
        'path': 'schemaName.WeaknessError',
        'types': list([
          dict({
            'canonical_path': 'schemaName.InvalidPokemonNumberError',
            'name': 'InvalidPokemonNumberError',
          }),
          dict({
            'canonical_path': 'schemaName.PokemonNotFoundError',
            'name': 'PokemonNotFoundError',
          }),
        ]),
      }),
      'WeaknessResponse': dict({
        'description': '',
        'fields': list([
          dict({
            'description': '',
            'name': 'data',
            'type': dict({
              'canonical_path': 'schemaName.Type',
              'is_list': True,
              'name': 'Type',
              'non_null': True,
              'non_null_list': False,
            }),
          }),
          dict({
            'description': '',
            'name': 'error',
            'type': dict({
              'canonical_path': 'schemaName.WeaknessError',
              'is_list': False,
              'name': 'WeaknessError',
              'non_null': False,
              'non_null_list': False,
            }),
          }),
        ]),
        'interfaces': list([
        ]),
        'name': 'WeaknessResponse',
        'path': 'schemaName.WeaknessResponse',
      }),
    }),
  })
# ---
